### DISCUSSION TITLE
Not sure how to load `.lua` files and their globals into the `Lua` struct

### DISCUSSION THREAD

**Discussion by TheLeonsver1** [mlua-rs/mlua#52] [Q&A] [answered]
I want to add user-modding to my game which is made with rust.
for that, I want to load player-made `.lua` files at runtime,
but I don't understand how it's done with `mlua`.

in the examples I see you loading a chunk, which if I understand correctly, is not a full source file but rather the insides of one lua function, and that's great, but I want to load a full source `.lua` file with its globals and functions from a string. and then call the global functions from that file but there are no examples of how to do that as far as I can tell.

my basic use-case for example would be to load a stringified `.lua` file into the `Lua` struct.
and then for example get the `main` function from the globals of the `Lua` struct and call that like you load `print` and call that in the `guided_tour` example.
would love to get your help, thank you!

**Comment by khvzak** [comment, marked as answer]
Loading stringified `.lua` file into Lua globals should work fine, eg:

```
fn main() {
    let lua = mlua::Lua::new();
    let source = std::fs::read_to_string("file.lua").unwrap();
    lua.load(&source).exec().unwrap();
    let globals = lua.globals();
    let main = globals.get::<_, mlua::Function>("main").unwrap();
    main.call::<_, ()>(()).unwrap();
}
```

file.lua:
```
function main()
    print("hello")
end
```

**Reply by khvzak** [reply]
Alternatively instead of globals you can also [set a specific environment](https://docs.rs/mlua/0.5.4/mlua/struct.Chunk.html#method.set_environment) for loaded file.

**Reply by TheLeonsver1** [reply]
Thank you very much for the help! I think you should maybe add this example to the guided tour example. it will help new users a lot I believe since I didn't think `lua.load(&source).exec().unwrap();` would work that way. Thank you very much again!

**Reply by zackfall** [reply]
I know this is an old discussion, but this is something that I can do with the luarocks plugins that I download? I mean, I want to do a neovim plugin, but I need to use plugins from luarocks. Can I do something like this:
```rust
fn main() {
    let lua = mlua::Lua::new();
    lua.load("conf = require(\"telescope.config\").values").exec()?;
    let globals = lua.globals();
    let conf = globals.get::<_, mlua::Table>("conf").unwrap();
}
```
and then use the values in it? or should I do something different in that case?

**Reply by khvzak** [reply]
@zackfall you example should work fine, `conf` is a global variable and will be available everywhere (with default env).
